### PR TITLE
Extend the plugin API by passing down a richer representation of the package graph

### PR DESCRIPTION
**Pull request by abertelrud**
Draft implementation of https://github.com/apple/swift-evolution/pull/1448

### Motivation:

The original `TargetBuildContext` structure passed to SwiftPM plugins was focused on build tools, which was the subject of SE-0303.  In order to prepare for new kinds of plugins such as custom commands (which have much more varied purposes and need more diverse information), we generalize this structure into a `PluginContext` that contains information about a whole package graph.

### Modifications:

The graph (being a DAG) is flattened and serialized into the plugin input JSON, and is reconstituted on the plugin side.  Path deduplication is used to reduce the amount of information sent over the wire.

A compatibility layer lets old plugins keep working with the old `TargetBuildContext`.

Note:  This code currently depends on synthesis of Codable for enums, so if it order to stay buildable with SwiftPM 5.4 and earlier we would need to add manually constructed Encoder/Decoder implementations.

### CODE REVIEW

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
I have realized that this PR depends on Swift 5.5's automatic synthesis of Codable conformance.  In order to keep SwiftPM buildable by 5.4, I'll amend this PR to add manual implementations of those methods.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

File: Sources/PackagePlugin/Context.swift [hunk 1]
```diff
@@ -6,14 +6,64 @@
...
+    /// any generated source files that should be processed further, and it
+    /// could include any caches used by the build tool or the plugin itself.
+    /// The plugin is in complete control of what is written under this di-
+    /// rectory, and the contents are preserved between builds.
+    ///
+    /// A plugin would usually create a separate subdirectory of this directory
+    /// for each command it creates, and the command would be configured to
+    /// write its outputs to that directory. The plugin may also create other
+    /// directories for cache files and other file system content that either
+    /// it or the command will need.
```
Comment (neonichu) [inline Sources/PackagePlugin/Context.swift:29]: Random question: what's the state of creating temporary directories in Swift? We have some code for this in `TSCBasic` and this being mentioned here made me wonder where we should offer some API to make that easier for plugin authors.

**Review by neonichu — APPROVED** [review]

File: Sources/PackagePlugin/Context.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/PackagePlugin/Context.swift:29]: That's a good point — you mean something like NSTemporaryDirectory(), or something inside this directory here?  I think the idea would be that the plugin should get the directories passed to it for both cache and temporary location.

Long-term I'd prefer if the plugin didn't access the file system at all, but rather used API provided by the plugin host.  That would also make it easier to make things more distributable.  But the need to shell out to local tools might actually spoil that.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Pull request merged**